lib/resourcecontrol/resource_control.go

```diff
@@ -0,0 +1,205 @@
+			writeAndDie(errPipe, err)
+		}
+
+		// TODO: drop privileges
+
+		unsetCustomEnvVars()
+
+		if len(os.Args) < 2 {
+			writeAndDie(errPipe, errNotEnoughArgs)
+		}
+
+		execPath, err := exec.LookPath(os.Args[1])
+		if err != nil {
+			writeAndDie(errPipe, err)
+		}
+
+		unix.CloseOnExec(errPipeFd)
+		if err := unix.Exec(execPath, os.Args[1:], os.Environ()); err != nil {
+			writeAndDie(errPipe, err)
+		}
+	}
+}
+
+func getCgroupRootPath() (string, error) {
+	r, err := os.Open("/proc/self/mounts")
+	if err != nil {
+		return "", err
+	}
+	defer r.Close()
+
+	var mountType, path, d string
+	for eof := false; !eof; {
+		_, err := fmt.Fscanf(r, "%s %s %s %s %s %s\n", &mountType, &path, &d, &d, &d, &d)
+
+		if err == io.EOF {
+			eof = true
+		} else if err != nil {
+			return "", err
+		}
+
+		if mountType == "cgroup2" {
+			return path, nil
+		}
+	}
+
+	// TODO: try to mount it before giving up
+
+	return "", errCgroupNotMounted
+}
+
+func getBlockDevs() ([]string, error) {
+	basePath := "/sys/block"
+
+	ds, err := os.ReadDir(basePath)
+	if err != nil {
+		return nil, err
+	}
+
+	var rs []string
+	for _, d := range ds {
+		dev, err := os.ReadFile(path.Join(basePath, d.Name(), "dev"))
+		if err != nil {
+			return nil, err
+		}
+		rs = append(rs, string(bytes.TrimSpace(dev)))
+	}
+
+	return rs, nil
+}
+
+func setResourceLimits() error {
+	cgroupPath, err := getCgroupRootPath()
+	if err != nil {
+		return err
+	}
+
+	cgroupPath = path.Join(cgroupPath, controlSubtree)
+
+	if err := os.Mkdir(cgroupPath, 0755); err != nil && !os.IsExist(err) {
+		return err
+	}
+
+	// Set CPU limits
+	if v, ok := os.LookupEnv(cpuMaxEnvVar); ok {
+		if err := os.WriteFile(path.Join(cgroupPath, "cpu.max"), []byte(v), 0700); err != nil {
+			return err
+		}
+	}
+
+	// Set memory limits
+	if v, ok := os.LookupEnv(memMaxEnvVar); ok {
+		if err := os.WriteFile(path.Join(cgroupPath, "memory.max"), []byte(v), 0700); err != nil {
+			return err
+		}
+	}
+
+	// Set IO limits
+	ioLimitString := ""
+	if v, ok := os.LookupEnv(ioMaxRbpsEnvVar); ok {
+		ioLimitString += " rbps=" + v
+	}
+
+	if v, ok := os.LookupEnv(ioMaxWbpsEnvVar); ok {
+		ioLimitString += " wbps=" + v
+	}
+
+	if ioLimitString != "" {
+		blkMajorMinors, err := getBlockDevs()
+		if err != nil {
+			return err
+		}
+
+		for _, b := range blkMajorMinors {
+			if err := os.WriteFile(path.Join(cgroupPath, "io.max"), []byte(b+ioLimitString), 0700); err != nil {
+				return err
+			}
+		}
+	}
+
+	// Add own PID to the cgroup
+	if err := os.WriteFile(path.Join(cgroupPath, "cgroup.procs"), []byte(strconv.Itoa(os.Getpid())), 0700); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Command takes the given name and args and returns a command with resource
+// limits enforced
+func Command(limits ResourceLimits, name string, args ...string) *Cmd {
+	cmd := exec.Command("/proc/self/exe", append([]string{name}, args...)...)
+	cmd.Env = append(os.Environ(), execEnvVar+"="+execEnvVar)
```

Comment:
passing all os.Env to the job process might not be totally safe.

Reply:
I agree, I think I left a comment mentioning that we should make them configurable.